state-sync/state-sync-v2/state-sync-driver/src/driver.rs

```diff
@@ -0,0 +1,297 @@
+    driver_configuration: DriverConfiguration,
+
+    // The event subscription service to notify listeners of on-chain events
+    event_subscription_service: EventSubscriptionService,
+
+    // The handler for notifications to mempool
+    mempool_notification_handler: MempoolNotificationHandler<M>,
+
+    // The storage synchronizer used to update local storage
+    storage_synchronizer: S,
+
+    // The client through which to stream data from the Diem network
+    streaming_service_client: StreamingServiceClient,
+}
+
+impl<M: MempoolNotificationSender, S: StorageSynchronizerInterface> StateSyncDriver<M, S> {
+    pub fn new(
+        bootstrap_notification_handler: BootstrapNotificationHandler,
+        consensus_notification_handler: ConsensusNotificationHandler,
+        driver_configuration: DriverConfiguration,
+        event_subscription_service: EventSubscriptionService,
+        mempool_notification_handler: MempoolNotificationHandler<M>,
+        storage_synchronizer: S,
+        streaming_service_client: StreamingServiceClient,
+    ) -> Self {
+        Self {
+            bootstrap_notification_handler,
+            consensus_notification_handler,
+            driver_configuration,
+            event_subscription_service,
+            mempool_notification_handler,
+            storage_synchronizer,
+            streaming_service_client,
+        }
+    }
+
+    /// Starts the state sync driver
+    pub async fn start_driver(mut self) {
+        let mut progress_check_interval =
+            IntervalStream::new(interval(Duration::from_millis(PROGRESS_CHECK_INTERVAL_MS))).fuse();
+
+        loop {
+            ::futures::select! {
+                notification = self.consensus_notification_handler.select_next_some() => {
+                    self.handle_consensus_notification(notification).await;
+                }
+                notification = self.bootstrap_notification_handler.select_next_some() => {
+                    self.handle_bootstrap_notification(notification);
+                },
+                _ = progress_check_interval.select_next_some() => {
+                    self.check_progress().await;
+                }
+            }
+        }
+    }
+
+    /// Handles a notification sent by consensus
+    async fn handle_consensus_notification(&mut self, notification: ConsensusNotification) {
+        // Handle the notification
+        let result = match notification {
+            ConsensusNotification::NotifyCommit(commit_notification) => {
+                self.handle_consensus_commit_notification(commit_notification)
+                    .await
+            }
+            ConsensusNotification::SyncToTarget(sync_notification) => {
+                self.handle_consensus_sync_notification(sync_notification)
+                    .await
+            }
+        };
+
+        // Log any errors from notification handling
+        if let Err(_error) = result {
+            // TODO(joshlind): log the error!
+        }
```

Comment:
😆 Yeah, this is partly laziness on my side: most of these `if let { //log error }` code blocks are copied across from state state v1, so I kept the code structure. To be honest, I don't feel that strongly about fixing them up, especially since the logging will come in a few days time 😄 If it's okay, I'm gonna follow them up in a future PR.